### ISSUE TITLE
Cody for JetBrain is consuming high CPU when running Cody Commands i.e. `Generate Test`

### ISSUE THREAD

**Issue by jay-fibi** [sourcegraph/jetbrains#509]
### Installation Information

Version:
5.2.21709

### Describe the bug

Steps to reproduce:
1. Have Cody plugin installed in your JetBrains IntelliJ IDEA editor
2. User is already logged in to Cody
3. Open any file and select a code snippet
4. Click on 'Cody' icon from right pane and go to 'Commands'
5. Click on 'Generate Test'
6. Open 'Activity Monitor' and search for IntelliJ and observe CPU usage during Cody action you have performed in step 4th and 5th

Observed behavior:
CPU usage is increased to ~50% during test generation.

### Expected behavior

CPU usage should not be spiked up very high.

### Additional context (logs, images, etc)

In the VS Code, if user perform same operation then CPU usage will not spiked up very high.

Video attachment:
https://drive.google.com/file/d/17ey32-g0WN7mX-O5jDrIp_5n2wk47xeo/view?usp=drive_link

**Comment by taylorsperry** [comment]
Thanks, @jay-fibi. We're aware of this issue and are working on a fix!

**Comment by exigow** [comment]
Looks like CPU usage increases when markdown appears in the chat, and it's not just about the situation where tokens are generated, but also when I simply scroll through the chat so this may be some kind of rendering issues. I'm looking for cause of these spikes.

**Comment by taylorsperry** [comment]
@jay-fibi would you please confirm that you're no longer seeing this issue on `v5.2.23880-nightly`?

**Comment by jay-fibi** [comment]
@taylorsperry I am still facing this issue with `v5.3.27-nightly`

**Comment by exigow** [comment]
My debugging so far:

1. Baseline: CPU usage on idle IDE is low around ~few percent.
2. Running the Generate Test command (or anything that contains Markdown in response) in native resolution (2560x1440 px) causes CPU usage spike from 2-5% to 10-20%.
3. After increasing the resolution 2x (to 5120x2880 px, virtually via `xrandr --output DP-0 --mode 2560x1440 --fb 5120x2880 --rate 165 --scale 2x2 --dpi 256`) the CPU usage increases significantly from 2-10% to 20-80% :rocket::rocket::rocket: for an identical command and response. The resolution increase is intended to simulate the resolution on Macs with hi-res displays. This also proves that the problem is with rendering performance (more pixels to draw -> higher CPU usage).
4. Profiler shows that there is a visible hot spot on AWT thread when response is generated. Looks like each generated token triggers the `CodeEditorPart.updateCode` method, which recognizes the language, initializes highlighter for this language and updates the content of the component. This code (recognize lang->overwrite highlighter->update content) is fired approx 600 times on each response. I suspect that each update causes repaint (I haven't verified that because we're creating virtual file with this code and we don't have control on that).
5. This code is definitely suboptimal: We could recognize the language and initialize the highlighter only once after the first successful recognition (instead doing this on each token). This may be tricky because it may give a false positive on the first tokens (for example, the first characters may suggest that this code is written in Python, and the next few characters suggest something different like Java).

To sum up: the problem is IMO not a blocker because it only occurs on higher resolutions that have a weaker CPU. However, we should definitely think how we can improve this suboptimal code.

I'm working on improving this.

-----

For the record, chart with spikes on high resolution:

![Selection_075](https://github.com/sourcegraph/jetbrains/assets/5013708/a673539f-5432-4677-a0a7-d39c0ffcb993)

On "normal" res:

![Selection_074](https://github.com/sourcegraph/jetbrains/assets/5013708/9be03069-b5ff-4884-81ce-7657faef0839)

Hot spot on AWT thread (yellow=our code):

![Selection_077](https://github.com/sourcegraph/jetbrains/assets/5013708/90a769fb-212a-4c17-bf8f-601bdf25829b)

**Comment by exigow** [comment]
@taylorsperry FYI

**Comment by taylorsperry** [comment]
Thank you @exigow , I'll bump this to P1

**Comment by mibali** [comment]
Hi @taylorsperry 
Another user reported that he is still seeing High CPU Spikes after upgrading to the latest version of the COdy Jetbrain Plugin. This is on the PhpStoem IDE.

**Comment by taylorsperry** [comment]
Thanks for checking in, @mibali! Is the CPU utilization causing the user's IDE to crash? Do you know if this issue only spikes when the user performs certain actions? And is this the [same user](https://sourcegraph.slack.com/archives/C059N5FRYG3/p1706721112641079) who has 16GB of RAM? (I ask because I think that sounds low, but I don't really know: my M1 has 64GB). 

@exigow is there any other information @mibali can share to help debug this?

**Comment by exigow** [comment]
> @exigow is there any other information @mibali can share to help debug this?

Sure! I have question: Does the high usage occur only during chat response generation (especially when code is rendered)? I need steps to reproduce this high CPU spikes. Please note that during inactivity, CPU usage should be minimal or almost non-existent. 

FYI I've made a fix for high CPU usage on Generate Test command: https://github.com/sourcegraph/jetbrains/pull/759

**Comment by exigow** [comment]
I identified more problems while playing around with the profiler:

- The chat window redraws excessively during scrolling and when tokens are generated (due to autoscroll). Scrolling causes frequent repaints of the AWT component of the chat list. This issue is especially noticeable at high resolutions.
- Chat responses allocate a large amount of memory (~500MB for a large response with code). This is likely caused by the initialization of a virtual code editor for each code block. High usage might also be due to excessive string allocation (when tokens are generated).
- Simple chat messages (those without code) are parsed as CommonMark, converted to an HTML string, and then parsed by AWT. It seems possible to streamline this process by converting Markdown directly into AWT components. Additionally, this MD->HTML->AWT loop is repeated for each token.

I'm looking on these issues.

**Comment by mibali** [comment]
@taylorsperry @exigow This time, the user isn't encountering high CPU usage but rather high RAM usage. This issue isn't limited to running commands in Cody; it persists when attempting to open PHPStorm after installing the Cody plugin. The entire PC crashes when trying to open any project in PHPStorm with Cody installed.

[Phpstorm+logs.zip](https://github.com/sourcegraph/jetbrains/files/14356254/Phpstorm%2Blogs.zip)

**Comment by exigow** [comment]
@mibali, thanks for the logs! I fixed one of the bugs that was causing excessive memory allocations that will likely make it into the upcoming release. 

There are still several issues to fix (listed in my previous messages) and I'm working on it. We're now mainly focused on bugs related to performance and stability.

**Comment by mibali** [comment]
@exigow When is the upcoming release?

**Comment by exigow** [comment]
Hi @mibali, release is ready! Currently, this release is **only available as nightly**, but as soon as we manage to test it with the help of Fibi-Labs, we will make it available as stable.

The exact list of changes is listed here: https://github.com/sourcegraph/jetbrains/releases/tag/v5.3.2625

This release contains few important fixes, including those related to memory/CPU usage that were the subject of this thread:

* https://github.com/sourcegraph/jetbrains/pull/759
* https://github.com/sourcegraph/jetbrains/pull/784
* https://github.com/sourcegraph/jetbrains/pull/804

I'm closing this issue, but if problems still persist or new ones related to performance are still present, please open them as separate issues. I'm asking for this mainly because the memory issue is a completely different problem from what was reported in this ticket and it's hard to track, even though it was also identified and addressed. Thanks in advance!

@jay-fibi FYI if problem with Generate Test still persists, please reopen this issue.
